### ISSUE TITLE
 lsp4e grinds eclipse to a halt

### ISSUE THREAD

**Issue by pauldaustin** [spring-projects/sts4#586]
See eclipse bug for an issue where lsp4e grinds eclipse to a halt.

https://bugs.eclipse.org/bugs/show_bug.cgi?id=570104

**Comment by martinlippert** [comment]
Thanks for raising the issue here as well, much appreciated. Given the thread dump you posted in the Eclipse report it would be extremely useful to capture a thread dump of the language server process while this timeout happens. You can find the process pretty easily using jps and looking for something that is called BootLanguagServerBootApp. Using jstack for that process will hopefully reveal more details.

**Comment by martinlippert** [comment]
@pauldaustin Did you have the chance to capture a thread dump for the language server?

**Comment by martinlippert** [comment]
Any news on this one here? Any chance to capture a thread dump of the spring-boot-language-server process while this happens or any steps how to reproduce this issue?

**Comment by pauldaustin** [comment]
Sorry I haven't had time to look into this in detail. I just turned it off as I don't need it.

My only thoughts are that the language server isn't starting at all, maybe a firewall issue.

My general thoughts are that Eclipse in the past has been great as it does everything in Java, so it's really fast. Now having experience with this and also in Visual Studio code where it launches external processes it seems slower and not as reliable.

**Comment by pcdavid** [comment]
I've been hit by this one several times this morning. It seems I get the following entries repeating in the workspace's error log:

```
!ENTRY org.springframework.tooling.ls.eclipse.commons 1 0 2021-03-15 10:44:54.616
!MESSAGE SpringBootLanguageServer exists!

!ENTRY org.springframework.tooling.ls.eclipse.commons 1 0 2021-03-15 10:44:55.642
!MESSAGE Boot LS startup time from start to initialized: 1026ms

!ENTRY org.eclipse.lsp4e 4 0 2021-03-15 10:44:55.672
!MESSAGE 
!STACK 0
java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1950)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2085)
	at org.eclipse.lsp4e.LanguageServerWrapper.supportsWorkspaceFolderCapability(LanguageServerWrapper.java:531)
	at org.eclipse.lsp4e.LanguageServerWrapper.canOperate(LanguageServerWrapper.java:520)
	at org.eclipse.lsp4e.LanguageServerWrapper.canOperate(LanguageServerWrapper.java:848)
	at org.eclipse.lsp4e.LanguageServiceAccessor.lambda$6(LanguageServiceAccessor.java:361)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1694)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSWrappers(LanguageServiceAccessor.java:362)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLanguageServers(LanguageServiceAccessor.java:588)
	at org.eclipse.lsp4e.operations.rename.LSPRenameHandler.setEnabled(LSPRenameHandler.java:121)
	at org.eclipse.ui.internal.handlers.HandlerProxy.setEnabled(HandlerProxy.java:229)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.canExecute(E4HandlerProxy.java:76)
	at jdk.internal.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:319)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:253)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:173)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.isEnabled(HandlerServiceHandler.java:62)
	at org.eclipse.core.commands.Command.isEnabled(Command.java:832)
	at org.eclipse.ui.menus.CommandContributionItem.isEnabled(CommandContributionItem.java:917)
	at org.eclipse.ui.menus.CommandContributionItem.updateMenuItem(CommandContributionItem.java:531)
	at org.eclipse.ui.menus.CommandContributionItem.update(CommandContributionItem.java:484)
	at org.eclipse.jface.action.MenuManager.update(MenuManager.java:860)
	at org.eclipse.jface.action.MenuManager.update(MenuManager.java:860)
	at org.eclipse.ui.internal.Workbench.updateActiveWorkbenchWindowMenuManager(Workbench.java:3152)
	at org.eclipse.ui.internal.Workbench.lambda$0(Workbench.java:3134)
	at org.eclipse.jface.bindings.BindingManager.fireBindingManagerChanged(BindingManager.java:903)
	at org.eclipse.jface.bindings.BindingManager.setActiveBindings(BindingManager.java:2183)
	at org.eclipse.jface.bindings.BindingManager.recomputeBindings(BindingManager.java:1744)
	at org.eclipse.jface.bindings.BindingManager.contextManagerChanged(BindingManager.java:693)
	at org.eclipse.core.commands.contexts.ContextManager.fireContextManagerChanged(ContextManager.java:164)
	at org.eclipse.core.commands.contexts.ContextManager.setEventCaching(ContextManager.java:323)
	at org.eclipse.core.commands.contexts.ContextManager.deferUpdates(ContextManager.java:85)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.activate(PartServiceImpl.java:777)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.activate(PartServiceImpl.java:679)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.activate(PartServiceImpl.java:674)
	at org.eclipse.ui.internal.WorkbenchPage.activate(WorkbenchPage.java:964)
	at org.eclipse.jdt.internal.ui.javaeditor.JavaOutlinePage$1.open(JavaOutlinePage.java:1021)
	at org.eclipse.ui.OpenAndLinkWithEditorHelper$InternalListener.open(OpenAndLinkWithEditorHelper.java:48)
	at org.eclipse.jface.viewers.StructuredViewer$2.run(StructuredViewer.java:802)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.viewers.StructuredViewer.fireOpen(StructuredViewer.java:799)
	at org.eclipse.jface.viewers.StructuredViewer.handleOpen(StructuredViewer.java:1118)
	at org.eclipse.jface.util.OpenStrategy.fireOpenEvent(OpenStrategy.java:275)
	at org.eclipse.jface.util.OpenStrategy.access$2(OpenStrategy.java:270)
	at org.eclipse.jface.util.OpenStrategy$1.handleEvent(OpenStrategy.java:310)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5785)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1427)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5048)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4526)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1157)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:653)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:590)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1461)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1434)

!ENTRY org.springframework.tooling.ls.eclipse.commons 1 0 2021-03-15 10:44:55.673
!MESSAGE Entering DelegatingStreamConnectionProvider()

!ENTRY org.springframework.tooling.ls.eclipse.commons 1 0 2021-03-15 10:44:55.673
!MESSAGE DelegatingStreamConnectionProvider classloader = org.eclipse.osgi.internal.loader.EquinoxClassLoader@68e7e3d4[org.springframework.tooling.boot.ls:1.24.0.202012131915(id=1750)]
```

This is using STS 4.9.0 on Linux (Ubuntun 20.10) with:

```
java.runtime.name=OpenJDK Runtime Environment
java.runtime.version=14.0.2+12-Ubuntu-1
```

I've restarted STS and the problem is gone for now. I'll try and take a thread dump if it occurs again.

Before I restarted, I could consistenly reproduce the issue when switching from a non-Java editor (in my case an EMF `.genmodel` file) to a Java editor. Switching between Java editors afterwards did not seem to trigger it.

**Comment by martinlippert** [comment]
@pcdavid It sounds to me like something is going wrong when starting up the Spring Boot language server, at least that is what I am guessing here. Can you enable the log output from the Spring Boot language server via `Preferences -> Language Servers -> Spring Language Servers` and then check the item for the `Spring Boot`. This will open a console view when the language server gets started and shows the log output of the language server process.

So the first question would be whether this output appears at all and if so, what shows up in there. Maybe there are some hints towards the underlying issue. Can you paste or upload that output when you hit this issue?

**Comment by pcdavid** [comment]
@martinlippert I've not forgotten about this, and will try and get a thread dump if I see it again, but I have not reproduced since my last comment.
BTW I'm updating to STS 4.10.0 right now.

**Comment by martinlippert** [comment]
@pcdavid Thanks for the update. Looking forward to more details once you updated to the latest release and in case you stumble upon this issue (or something related) again. Would be great to get more details to see what might be causing this under the hood to get this fixed and improved.

**Comment by pcdavid** [comment]
FWIW, since I've upgraded to STS 4.10.0 I've never experienced it again.

**Comment by martinlippert** [comment]
@pcdavid Good news, thanks for sharing. Looks like the various changes that we implemented for the 4.10.0 release are paying off here.

With that, let me close this issue for now. If you or someone else come across this again, please comment and we can re-open this.
